[pkg/ottl] Support for append (#33017)
open-telemetry/opentelemetry-collector-contrib: pkg/ottl/ottlfuncs/README.md

- [set](#set)
- [truncate_all](#truncate_all)

### append

> `append(target, Optional[value], Optional[values])`

Reviewer: Lets add a warning in this function's docs that `append` can be used to create heterogeneus slices, and that otel attributes don't support that.

> The `append` function appens single or multiple string values to target. 

Reviewer 2: ```suggestion
The `append` function appends single or multiple string values to `target`. 
```

> `append` converts scalar values into an array if the field exists but is not an array and creates an array containing the provided values if the field doesn’t exist.

Reviewer 2: ```suggestion
`append` converts scalar values into an array if the field exists but is not an array, and creates an array containing the provided values if the field doesn’t exist.
```
I feel like we need a break somewhere here.

> Resulting field is always of type `pcommon.Slice` keeping types of existing values appending values as set not performing any conversions.

Reviewer 2: ```suggestion
Resulting field is always of type `pcommon.Slice` and will not convert the types of existing or new items in the slice.
```

- [set](#set)
- [truncate_all](#truncate_all)

### append

`append(target, Optional[value], Optional[values])`

The `append` function appends single or multiple string values to `target`. 
`append` converts scalar values into an array if the field exists but is not an array, and creates an array containing the provided values if the field doesn’t exist.

> Resulting field is always of type `pcommon.Slice` and will not convert the types of existing or new items in the slice. This means that it is possible to create a slice whose elements have different types.  Be careful when using `append` to set attribute values, as OTLP does not support slices as attribute values when the elements in the slice are not all the same type.

Reviewer 2: ```suggestion
Resulting field is always of type `pcommon.Slice` and will not convert the types of existing or new items in the slice. This means that it is possible to create a slice whose elements have different types.  Be careful when using `append` to set attribute values, as this will produce values that are not possible to create through OpenTelemetry APIs [according to](https://opentelemetry.io/docs/specs/otel/common/#attribute) the OpenTelemetry specification.
```

>OTLP does not support slices as attribute values when the elements in the slice are not all the same type.

I don't think this is true since the proto supports this (take [log](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/logs/v1/logs.proto#L175) or [resource](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/resource/v1/resource.proto#L32) attributes for example). The only reference I see to this is [in the spec](https://opentelemetry.io/docs/specs/otel/common/#attribute) which we shouldn't limit ourselves to.

I have a suggesting for the wording I think we should take here.

  - `append(attributes["tags"], "prod")`
  - `append(attributes["tags"], values = ["staging", "staging:east"])`
  - `append(attributes["tags_copy"], attributes["tags"])`

> 
> `append` is able to create heterogenous slices that are not supported by Open Telemetry attributes.

Reviewer 2: ```suggestion
```

We have this addressed above now, so we should try to only mention it once. I do think it is worth mentioning earlier.

`append(target, Optional[value], Optional[values])`

The `append` function appends single or multiple string values to `target`. 
`append` converts scalar values into an array if the field exists but is not an array, and creates an array containing the provided values if the field doesn’t exist.

> Resulting field is always of type `pcommon.Slice` and will not convert the types of existing or new items in the slice.

Reviewer: ```suggestion
Resulting field is always of type `pcommon.Slice` and will not convert the types of existing or new items in the slice. This means that it is possible to create a slice whose elements have different types.  Be careful when using `append` to set attribute values, as OTLP does not support slices as attribute values when the elements in the slice are not all the same type.
```
